server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java

```diff
@@ -479,10 +482,15 @@ private static int component(final Edge edge, final int id, final ArrayList<Edge
      * @param coordinates Array of coordinates to write the result to
      * @return the coordinates parameter
      */
-    private static Coordinate[] coordinates(Edge component, Coordinate[] coordinates) {
+    private static Coordinate[] coordinates(Edge component, Coordinate[] coordinates, double[] partitionPoint) {
         for (int i = 0; i < coordinates.length; i++) {
             coordinates[i] = (component = component.next).coordinate;
         }
+        // First and last coordinates must be equal
+        if (coordinates[0].equals(coordinates[coordinates.length - 1]) == false) {
+            throw new InvalidShapeException("Self-intersection at or near point ["
+                + partitionPoint[0] + "," + partitionPoint[1] + "," + partitionPoint[2] + "]");
```

Comment:
I would add a check for `partitionPoint[2]` value and only print it if it is not `NULL_ORDINATE` (`NaN`).